File: middle-linked-list.php
<?php
/**
 * Definition for a singly-linked list.
 */
class ListNode {
     public $val = 0;
     public $next = null;
     function __construct($val = 0, $next = null) {
         $this->val = $val;
         $this->next = $next;
     }
}

class Solution {
    var $head;
    function __construct()
    {
        $this->head = null;
    }

    function push($data)
    {
        $new_node = new ListNode();
        $new_node->next = $this->head;
        $new_node->val = $data;
        $this->head = $new_node;
    }

    function printList()
    {
        $tmp = $this->head;
        $k = 0;
        while($tmp)
        {
            echo $tmp->val.' ';
            $tmp = $tmp->next;
        }
    }
    /**
     * @param ListNode $head
     * @return ListNode
     */
    function middleNode($head) 
    {
        $slow = $fast = $head;
        while( $fast && $fast->next )
        {
            $slow = $slow->next;
            $fast = $fast->next->next;
        }
        return $slow;
    }
}

$sol = new Solution();
$sol->push(10);
$sol->push(20);
$sol->push(30);
$sol->push(40);
$sol->push(50);
$sol->push(60);
$sol->push(70);
echo "here";
$sol->printList();
echo "\nhere:";
var_dump($sol->middleNode($sol->head));

File: middle-linked-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def __init__(self):
        self.head = None

    def push(self, data):
        new_node = ListNode()
        new_node.val = date
        new_node.next = self.head
        self.head = new_node

    def printList(self):
        tmp = self.head
        while(tmp):
            print(tmp.val, end=" ")
            tmp = tmp.next

    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

sol = Solution()
sol.push(10)
sol.push(20)
sol.push(30)
sol.push(40)
sol.push(50)
sol.push(60)
sol.printList()
s = sol.middleNode(sol.head)
print(s)
